zerver/tests/test_user_groups.py

```diff
@@ -361,42 +362,88 @@ def test_update_members_of_user_group(self) -> None:
+        )
+        self.assertEqual(self.get_last_message().content, expected_notification)
+
         # Test adding a member already there.
         result = self.client_post(f"/json/user_groups/{user_group.id}/members", info=params)
         self.assert_json_error(result, f"User {othello.id} is already a member of this group")
         self.assertEqual(UserGroupMembership.objects.filter(user_group=user_group).count(), 2)
         members = get_direct_memberships_of_users(user_group, [hamlet, othello])
         self.assert_length(members, 2)
 
+        # Test user adding itself to user group
+        desdemona = self.example_user("desdemona")
+        self.login_user(desdemona)
+
+        params = {"add": orjson.dumps([desdemona.id]).decode()}
+        initial_last_message = self.get_last_message()
+        result = self.client_post(f"/json/user_groups/{user_group.id}/members", info=params)
+        self.assert_json_success(result)
+
+        self.assertEqual(UserGroupMembership.objects.filter(user_group=user_group).count(), 3)
+        members = get_direct_memberships_of_users(user_group, [hamlet, othello, desdemona])
+        self.assert_length(members, 3)
+
+        # No notification message is sent for adding to user group
+        self.assertEqual(self.get_last_message(), initial_last_message)
+
         aaron = self.example_user("aaron")
 
         # For normal testing we again log in with hamlet
         self.logout()
         self.login_user(hamlet)
         # Test remove members
         params = {"delete": orjson.dumps([othello.id]).decode()}
+        initial_last_message = self.get_last_message()
         result = self.client_post(f"/json/user_groups/{user_group.id}/members", info=params)
         self.assert_json_success(result)
-        self.assertEqual(UserGroupMembership.objects.filter(user_group=user_group).count(), 1)
-        members = get_direct_memberships_of_users(user_group, [hamlet, othello, aaron])
-        self.assert_length(members, 1)
+
+        self.assertEqual(UserGroupMembership.objects.filter(user_group=user_group).count(), 2)
+        members = get_direct_memberships_of_users(user_group, [hamlet, othello, aaron, desdemona])
+        self.assert_length(members, 2)
+
+        # A notification message is sent for removing from user group
+        self.assertNotEqual(self.get_last_message(), initial_last_message)
+        expected_notification = (
+            f"{silent_mention_syntax_for_user(hamlet)} unsubscribed you from the group @_*support*."
+        )
+        self.assertEqual(self.get_last_message().content, expected_notification)
 
         # Test remove a member that's already removed
         params = {"delete": orjson.dumps([othello.id]).decode()}
         result = self.client_post(f"/json/user_groups/{user_group.id}/members", info=params)
         self.assert_json_error(result, f"There is no member '{othello.id}' in this user group")
+
+        # Test user delete itself from user group
```

Comment:
For example, you could change this test to remove 3 users: (bot, self, deactivated user) and cover all 3 of those corner cases together in a single API request, just adding some extra assertions at the end.

(And of course the previous similar one to add 3 users).

Reply:
I have added the test as per your comments.